File: Mage/CoreData/StaticLayer.m
//
//  StaticLayer.m
//  mage-ios-sdk
//
//  Created by William Newman on 4/13/16.
//  Copyright © 2016 National Geospatial-Intelligence Agency. All rights reserved.
//

#import "StaticLayer.h"

#import "MageSessionManager.h"
#import "MageServer.h"

@implementation StaticLayer

NSString * const StaticLayerLoaded = @"mil.nga.giat.mage.static.layer.loaded";

+ (NSString *) layerIdFromJson:(NSDictionary *) json {
    return [json objectForKey:@"id"];
}

+ (void) createOrUpdateStaticLayer: (id) layer withEventId: (NSNumber *) eventId inContext: (NSManagedObjectContext *) context {
    NSString *remoteLayerId = [StaticLayer layerIdFromJson:layer];
    StaticLayer *l = [StaticLayer MR_findFirstWithPredicate:[NSPredicate predicateWithFormat:@"(remoteId == %@ AND eventId == %@)", remoteLayerId, eventId] inContext:context];
    if (l == nil) {
        l = [StaticLayer MR_createEntityInContext:context];
        [l populateObjectFromJson:layer withEventId:eventId];
        l.loaded = [NSNumber numberWithFloat: OFFLINE_LAYER_NOT_DOWNLOADED];
        NSLog(@"Inserting layer with id: %@ into event: %@", l.remoteId, eventId);
    } else {
        NSLog(@"Updating layer with id: %@ into event: %@", l.remoteId, eventId);
        [l populateObjectFromJson:layer withEventId:eventId];
    }
    NSLog(@"layer loaded %@ ? %@", l.name, l.loaded);
    if (l.loaded) {
        [StaticLayer fetchStaticLayerData:eventId layer:l];
    }
}

+ (void) fetchStaticLayerData: (NSNumber *)eventId layer: (StaticLayer *) staticLayer {
    MageSessionManager *manager = [MageSessionManager sharedManager];
    NSURLSessionDataTask *fetchFeaturesTask = [StaticLayer operationToFetchStaticLayerData:staticLayer];
    [manager addTask:fetchFeaturesTask];
}

+ (NSURLSessionDataTask *) operationToFetchStaticLayerData: (StaticLayer *) layer {
    MageSessionManager *manager = [MageSessionManager sharedManager];
    __block NSNumber *layerId = layer.remoteId;
    __block NSNumber *eventId = layer.eventId;
    
    // put this line back when the event endpoint returns the proper url for the layer
    //NSURL *URL = [NSURL URLWithString:[NSString stringWithFormat:@"%@/features", layer.url]];
    NSURL *URL = [NSURL URLWithString:[NSString stringWithFormat:@"%@/api/events/%@/layers/%@/features", [MageServer baseURL], eventId, layerId]];
    NSURLSessionDataTask * task = [manager GET_TASK:URL.absoluteString parameters:nil progress:nil success:^(NSURLSessionTask *task, id responseObject) {
        StaticLayer *fetchedLayer = [StaticLayer MR_findFirstWithPredicate:[NSPredicate predicateWithFormat:@"remoteId == %@ AND eventId == %@", layerId, eventId]];
        [MagicalRecord saveWithBlock:^(NSManagedObjectContext *localContext) {
            StaticLayer *localLayer = [fetchedLayer MR_inContext:localContext];
            NSLog(@"fetched static features for %@", localLayer.name);
            NSMutableDictionary *dictionaryResponse = (NSMutableDictionary *)CFBridgingRelease(CFPropertyListCreateDeepCopy(kCFAllocatorDefault, (CFDictionaryRef)responseObject, kCFPropertyListMutableContainers));
            
            for (NSDictionary *feature in [dictionaryResponse objectForKey:@"features"]) {
                NSString *iconUrl = [feature valueForKeyPath:@"properties.style.iconStyle.icon.href"];
                if (iconUrl && [iconUrl hasPrefix:@"https"]) {
                    NSString *documentsDirectory = [NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES)objectAtIndex:0];
                    NSString *featureIconRelativePath = [NSString stringWithFormat:@"featureIcons/%@/%@", localLayer.remoteId, [feature valueForKey:@"id"]];
                    NSString *featureIconPath = [NSString stringWithFormat:@"%@/%@", documentsDirectory, featureIconRelativePath];
                    NSData *imageData = [NSData dataWithContentsOfURL:[NSURL URLWithString:iconUrl]];
                    
                    NSError *error;
                    if (![[NSFileManager defaultManager] fileExistsAtPath:featureIconPath])
                        [[NSFileManager defaultManager] createDirectoryAtPath:[featureIconPath stringByDeletingLastPathComponent] withIntermediateDirectories:YES attributes:nil error:&error];
                    
                    [imageData writeToFile:featureIconPath atomically:YES];
                    [feature setValue:featureIconRelativePath forKeyPath:@"properties.style.iconStyle.icon.href"];
                }
            }
            localLayer.data = dictionaryResponse;
            localLayer.loaded = [NSNumber numberWithFloat: OFFLINE_LAYER_LOADED];
            localLayer.downloading = nil;
            
        } completion:^(BOOL contextDidSave, NSError *error) {
            if (contextDidSave) {
                StaticLayer *localLayer = [layer MR_inContext:[NSManagedObjectContext MR_defaultContext]];
                [[NSNotificationCenter defaultCenter] postNotificationName:StaticLayerLoaded object:localLayer];
            }
        }];
    } failure:^(NSURLSessionTask *operation, NSError *error) {
        NSLog(@"Error: %@", error);
    }];
    [MagicalRecord saveWithBlock:^(NSManagedObjectContext *localContext) {
        Layer *localLayer = [layer MR_inContext:localContext];
        
        localLayer.downloading = YES;
    }];
    
    return task;
}

- (void) removeStaticLayerData {
    __weak typeof(self) weakSelf = self;
    [MagicalRecord saveWithBlock:^(NSManagedObjectContext *localContext) {
        StaticLayer *localLayer = [weakSelf MR_inContext:localContext];
        localLayer.loaded = [NSNumber numberWithFloat: OFFLINE_LAYER_NOT_DOWNLOADED];
        localLayer.data = nil;
    } completion:^(BOOL contextDidSave, NSError *error) {
   }];
}

@end

File: Mage/CoreData/StaticLayer.swift
//
//  StaticLayer.m
//  mage-ios-sdk
//
//  Created by William Newman on 4/13/16.
//  Copyright © 2016 National Geospatial-Intelligence Agency. All rights reserved.
//

import Foundation
import CoreData
import ServerDTO
import Layer
import Persistence

extension StaticLayer {
    
    public var features: [[AnyHashable: Any]]? {
        get {
            return data?["features"] as? [[AnyHashable: Any]]
        }
    }
    
    static func featureName(feature: [AnyHashable : Any]) -> String? {
        return (feature["properties"] as? [AnyHashable : Any])?["name"] as? String
    }
    
    static func featureDescription(feature: [AnyHashable : Any]) -> String? {
        return (feature["properties"] as? [AnyHashable : Any])?["description"] as? String
    }
    
    static func featureTimestamp(feature: [AnyHashable : Any]) -> Date? {
        guard let timestamp = (feature["properties"] as? [AnyHashable : Any])?["timestamp"] as? String else {
            return nil
        }
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withDashSeparatorInDate, .withFullDate, .withTime, .withColonSeparatorInTime, .withTimeZone];
        formatter.timeZone = TimeZone(secondsFromGMT: 0)!;
        let lastModifiedDate = formatter.date(from: timestamp) ?? Date();
        return lastModifiedDate
    }
    
    static func featureType(feature: [AnyHashable : Any]) -> String? {
        return (feature["geometry"] as? [AnyHashable : Any])?["type"] as? String
    }
    
    static func featureCoordinates(feature: [AnyHashable : Any]) -> [Any]? {
        return (feature["geometry"] as? [AnyHashable : Any])?["coordinates"] as? [Any]
    }
    
    static func featureFillOpacity(feature: [AnyHashable : Any]) -> Double {
        return (feature as NSDictionary).value(forKeyPath: "properties.style.polyStyle.color.opacity") as? Double ?? 255.0
    }
    
    static func featureFillColor(feature: [AnyHashable : Any]) -> String {
        return (feature as NSDictionary).value(forKeyPath: "properties.style.polyStyle.color.rgb") as? String ?? "#000000"
    }
    
    func featureLineOpacity(feature: [AnyHashable : Any]) -> Double {
        return (feature as NSDictionary).value(forKeyPath: "properties.style.lineStyle.color.opacity") as? Double ?? 255.0
    }
    
    static func featureLineColor(feature: [AnyHashable : Any]) -> String {
        return (feature as NSDictionary).value(forKeyPath: "properties.style.lineStyle.color.rgb") as? String ?? "#000000"
    }
    
    static func featureLineWidth(feature: [AnyHashable : Any]) -> Double {
        let width = (feature as NSDictionary).value(forKeyPath: "properties.style.lineStyle.width")
        if let doubleWidth = width as? Double {
            return doubleWidth
        } else if let stringWidth = width as? String {
            return Double(stringWidth) ?? 1.0
        }
        return 1.0
    }
    
    @objc public static let StaticLayerLoaded = "mil.nga.giat.mage.static.layer.loaded";
    
    @objc public static func createOrUpdate(json: [AnyHashable : Any], eventId: NSNumber, context: NSManagedObjectContext) {
        guard let remoteLayerId = Layer.layerId(json: json) else {
            return;
        }
        
        var l = StaticLayer.mr_findFirst(with: NSPredicate(format:"(\(LayerKey.remoteId.key) == %@ AND \(LayerKey.eventId.key) == %@)", remoteLayerId, eventId), in: context);
        if l == nil {
            l = StaticLayer.mr_createEntity(in: context);
            l?.populate(json, eventId: eventId);
            l?.loaded = NSNumber(floatLiteral: OFFLINE_LAYER_NOT_DOWNLOADED);
            NSLog("Inserting layer with id: \(l?.remoteId ?? -1) into event \(eventId)")
        } else {
            NSLog("Updating layer with id: \(l?.remoteId ?? -1) into event \(eventId)")
            l?.populate(json, eventId: eventId);
        }
        guard let l = l else {
            return;
        }
        NSLog("layer loaded \(l.name ?? "unkonwn")? \(l.loaded ?? -1.0)")
    }
    
    @objc public static func fetchStaticLayerData(eventId: NSNumber, staticLayer: StaticLayer) {
        Task {
            try await DependencyContainer.shared.useCaseFactory
                .resolve(.StaticLayerDataFetchUseCase)
                .execute(eventID: EventID(eventId), layerID: staticLayer.remoteId.map(LayerID.init))
        }
    }
    
    @objc public func removeStaticLayerData() {
        MagicalRecord.save { [weak self] context in
            guard let localLayer = self?.mr_(in: context) else {
                return;
            }
            localLayer.loaded = NSNumber(floatLiteral: Layer.OFFLINE_LAYER_NOT_DOWNLOADED);
            localLayer.data = nil
        } completion: { contextDidSave, error in
        }
    }
}
